src/log.rs

```diff
@@ -0,0 +1,467 @@
+    ) {
+        let log_domain: GString = from_glib_borrow(log_domain);
+        let message: GString = from_glib_borrow(message);
+        let callback: &P = &*(user_data as *mut _);
+        (*callback)(
+            log_domain.as_str(),
+            from_glib(log_level),
+            from_glib(log_level),
+            message.as_str(),
+        );
+    }
+    let log_func = Some(log_func_func::<P> as _);
+    unsafe extern "C" fn destroy_func<
+        P: Fn(&str, LogLevel, LogFlags, &str) + Send + Sync + 'static,
+    >(
+        data: glib_sys::gpointer,
+    ) {
+        let _callback: Box_<P> = Box_::from_raw(data as *mut _);
+    }
+    let destroy_call4 = Some(destroy_func::<P> as _);
+    let super_callback0: Box_<P> = log_func_data;
+    unsafe {
+        from_glib(glib_sys::g_log_set_handler_full(
+            log_domain.to_glib_none().0,
+            log_levels.to_glib() | log_flags.to_glib(),
+            log_func,
+            Box_::into_raw(super_callback0) as *mut _,
+            destroy_call4,
+        ))
+    }
+}
+
+pub fn log_remove_handler(log_domain: &str, handler_id: LogHandlerId) {
+    unsafe {
+        glib_sys::g_log_remove_handler(log_domain.to_glib_none().0, handler_id.to_glib());
+    }
+}
+
+pub fn log_set_always_fatal(fatal_levels: LogLevels) -> LogLevels {
+    unsafe { from_glib(glib_sys::g_log_set_always_fatal(fatal_levels.to_glib())) }
+}
+
+pub fn log_set_fatal_mask(log_domain: &str, fatal_levels: LogLevels) -> LogLevels {
+    unsafe {
+        from_glib(glib_sys::g_log_set_fatal_mask(
+            log_domain.to_glib_none().0,
+            fatal_levels.to_glib(),
+        ))
+    }
+}
+
+// #[cfg(any(feature = "v2_50", feature = "dox"))]
+// pub fn log_variant(log_domain: Option<&str>, log_level: LogLevel, fields: &Variant) {
+//     unsafe {
+//         glib_sys::g_log_variant(
+//             log_domain.to_glib_none().0,
+//             log_level.to_glib(),
+//             fields.to_glib_none().0,
+//         );
+//     }
+// }
+
+static PRINT_HANDLER: Lazy<Mutex<Option<Arc<Box_<Box_<dyn Fn(&str) + Send + Sync + 'static>>>>>> =
+    Lazy::new(|| Mutex::new(None));
+
+/// To set back the default print handler, use the [`unset_print_handler`] function.
+pub fn set_print_handler<P: Fn(&str) + Send + Sync + 'static>(func: P) {
+    unsafe extern "C" fn func_func(string: *const libc::c_char) {
+        if let Some(callback) = match PRINT_HANDLER.lock() {
+            Ok(handler) => {
+                if let Some(ref handler) = *handler {
+                    Some(Arc::clone(handler))
+                } else {
+                    panic!("PRINT_HANDLER cannot be None!");
+                }
+            }
+            Err(_) => {
+                // should we log something here?
+                None
+            }
+        } {
+            let string: GString = from_glib_borrow(string);
+            (*callback)(string.as_str())
+        }
+    }
+    let func = Some(Arc::new(Box_::new(Box_::new(func))));
```

Comment:
Make sure to give this the expected type, because below you do a blanket transmute and otherwise this is a bit dangerous.